### DISCUSSION TITLE
Is this pattern already implemented somewhere?

### DISCUSSION THREAD

**Discussion by davisyoshida** [jax-ml/jax#16259] [Q&A] [answered]
I noticed that in writing both a 4-bit matrix implementation and a LoRA implementation that there's a common pattern of "thing which represents a matrix but avoids instantiating a dense matrix if possible." For example with 4-bit quantization you prefer to use a specialized matmul kernel, but if you need to do something like calculate the norm of a matrix you can always full materialize it.

It's pretty easy to implement this pattern with tracers, but handling some of the stuff like dealing with pjit and remat primitives every time is annoying. Is this special case already implemented somewhere? For example I know there's a `__jax_array__` API, so perhaps it's possible to implement it that way instead of needing to reinvent the wheel.

**Comment by jakevdp** [comment, marked as answer]
No, I don't believe there's any stock pattern for this kind of thing. One note on `__jax_array__` is that it's not a fully-implemented API (and probably never will be, see #4725), so I would not recommend relying on it (cc/ @mattjj, who knows more about the plans for this).

**Reply by davisyoshida** [reply]
Cool I'll just try to get this one to a good enough state for my own use then. One question: I tried to implement handling of pjit as follows:

```python
        elif primitive.name == 'pjit':    
            if params['resource_env'] is None:    
                closed_jaxpr = params['jaxpr']
                outs = wrap_jaxpr(    
                    fn=partial(    
                        pjit.pjit,    
                        donate_argnums=tuple(i for i, donate in enumerate(params['donated_invars'], 1) if donate),    
                        keep_unused=params['keep_unused'],    
                        inline=params['inline'],    
                    ),    
                    jaxpr=closed_jaxpr.jaxpr,    
                    literals=closed_jaxpr.literals,    
                    vals=vals    
                )
             else:    
                warnings.warn(f'Encountered pjit with non-None resource_env, so implict args will be materialized.'
                ...

def wrap_jaxpr(fn, jaxpr, literals, vals):               
    def new_fn(consts, *args):    
        return core.eval_jaxpr(jaxpr, consts, *args)    
    
    wrapped_fn = lu.wrap_init(use_implicit_args(new_fn))    
    
    flat_args, in_tree = jax.tree_util.tree_flatten((literals, *vals))    
    flat_fn, out_tree = flatten_fun_nokwargs(wrapped_fn, in_tree)    
    
    result = fn(flat_fn.call_wrapped)(*flat_args)    
    unflattened = jax.tree_util.tree_unflatten(out_tree(), result)    
    return unflattened
```

1. I'm not sure if I'm using the `linear_util` stuff correctly, what precisely should I be wrapping? Should I somehow be making my whole trace thing into a linear_util transformation?
2. I'm not sure how to properly forward the pjit primitive params, since they don't line seem to line up 1-to-1 with the pjit function arguments. Are there any examples I can refer to for that?

**Reply by jakevdp** [reply]
Is this in the context of writing a jaxpr interpreter?

**Reply by davisyoshida** [reply]
This is a trace based one. The `elif` block is from the `process_primitive` on the Trace.

Edit:
Oh I guess I cut out a pretty important line which is:
```python
vals = [t.value for t in tracers]
```
So `vals` is just the unboxed values from the tracers the primitive is being called on.

**Reply by jakevdp** [reply]
Sorry, I'm having trouble extrapolating the context of your code from the snippet you provided. What is `pass_through_to_jaxpr`?

**Reply by davisyoshida** [reply]
Ah sorry about that, I renamed that to `wrap_jaxpr` but missed changing it in that spot. Currently I'm using `wrap_jaxpr` to make a recursive call to my transformation then re-apply pjit (I do the same for remat2). Basically the issue I'm having is that if I get some pjit params such as:

```python
{'jaxpr': { lambda ; a:f32[2,3] b:f32[2,3]. let
    c:f32[2,3] = mul a b
    d:f32[] = reduce_sum[axes=(0, 1)] c
  in (d,) },
'in_shardings': (UnspecifiedValue, UnspecifiedValue), 'out_shardings': (UnspecifiedValue,), 'resource_env': ResourceEnv(Mesh(device_ids=[], axis_names=()), ()), 'donated_invars': (False, False), 'name': 'f', 'keep_unused': False, 'inline': False}
```

I'm not sure how to create a new pjit call which is identical other than the jaxpr being wrapped. This is probably a more self-contained way of asking the question since I don't think it depends too much on the particulars of my implementation.

**Reply by jakevdp** [reply]
I think rather than doing things manually, what you want is to call `get_bind_params` and then bind the original primitive, something like this:
```python
subfuns, bind_params = primitive.get_bind_params(params)
result = primitive.bind(*subfuns, *vars, **bind_params)
```

**Reply by davisyoshida** [reply]
That's what I do when I want to execute a primitive with no modification, but in this case I need to at the very least replace the jaxpr which the pjit primitive is executing, and possibly do more things as well if the number of invars is changing (e.g. for a quantized matrix each original input might get transformed to 3 inputs, an int matrix + zero + scale).

Would the following be an okay strategy?
1) Get the bind params
2) Transform the original jaxpr using my transformation and trace its evaluation into a new jaxpr
3) Set the 'jaxpr' field in bind params equal to the new jaxpr
4) Take a guess at how to update 'in_shardings'/'out_shardings' such as duplicating entries in them based on the original input -> transformed inputs mapping.

I'm currently not even doing any sharding, it just happens that pjit primitives get emitted a lot and I want to make sure my code will handle them correctly.

**Reply by jakevdp** [reply]
Yeah, that sounds reasonable. I'm not sure how to correctly handle shardings in this case – shardings are a new enough concept (and custom final-style transformations are advanced enough) that, without exaggeration, I'd guess there's probably nobody who has experience doing this yet!

**Reply by davisyoshida** [reply]
Alright I'll give that a shot. Thanks a ton for the quick responses.

**Reply by davisyoshida** [reply]
"Easy to implement this pattern with tracers" may have been premature :sweat_smile: . Just in case anyone comes across this in the future, it seems like the "transform the jaxpr" approach works. The main thing to be careful of is accidentally transforming a jaxpr with no literals into one that has literals. This matters for `remat2` since it never passes any literals to its jaxpr. The `jax.interpreters.partial_eval.convert_constvars_jaxpr` function can rectify this though.

**Comment by patrick-kidger** [comment]
I have two other cases where this would be a useful feature to have -- sparse arrays, and symbolic zeros. (The latter outside of AD, as a general jaxtype.)

I note that sparse arrays already exist and have some limited support. So would it be possible to build into JAX a single unified approach here?

**Reply by patrick-kidger** [reply]
And actually I just saw #16289, which is another example of this use-case.

**Reply by davisyoshida** [reply]
Oh that's interesting I was using a symbolic zero for some unit testing, but I didn't think it had much actual use. What's the use case for it?

**Reply by patrick-kidger** [reply]
The main one I have is prior to flattening a pytree into a vector or matrix, for passing to linear algebra routines. Often we can just skip over anything that's zero as we know how the linear algebra routines will behave on that, and so we can reduce our computational cost.

**Reply by davisyoshida** [reply]
I think one question is how to make it so that functions execute different behavior when passed one of these objects though. The angle I was coming from was trying to avoid that happening. I suppose you could check "hey is this a Zero/4Bit/EtcTracer" but that feels a bit hacky.

**Reply by patrick-kidger** [reply]
Have it be reflected in the aval. (It should be done in the aval, not the tracer, so that e.g. I can wrap a batch tracer around a symbolic zero.) Then have Primitives perform multiple dispatch depending on the aval of their arguments.

**Reply by davisyoshida** [reply]
Is that something that would require modification of JAX internals? I'm not actually clear on what all the legal aval types are.

**Reply by patrick-kidger** [reply]
I think JAX has most of the notions for this already. We can already write custom primitives and it has avals, etc. The main things are (a) creating custom jaxtypes+their avals is still largely unexplored, and (b) ideally primitives would support multiple dispatch a bit more neatly than having us copy-paste it between their abstract/impl/lowering rules.

**Reply by davisyoshida** [reply]
Ah yeah it looks like all my tests pass if I make a custom aval inherits from ShapedArray. Cool! 

I think to make it fully work properly I need to at least update `jax.core.raise_to_shaped_mappings` and probably some other places I'm not aware of.

As for the copy-pasting thing, aren't tracers intended for things like this? You can choose to change how a primitive executes, or defer to the default handler.

**Reply by patrick-kidger** [reply]
So JAX plays pretty fast-and-loose with stuff like inheritance. (And mostly does stuff like `if type(x) is Foo` rather than `isinstance(x, Foo)`. It might be safer to create an entirely new aval.

Indeed I think there are probably multiple different registries that need to be told how to handle the new aval.

---

On tracers: hmm, maybe that is another option! Don't introduce another aval, but instead introduce a `MultipleDispatchTrace` with corresponding `MultipleDispatchTracer`. Aval-wise this is still just a a `ShapedArray`, but it carries additional metadata that the rule used to dispatch appropriately.

I think I like that! It's basically a generalisation of the existing `jax.experimental.sparse.sparsify` transform, or of your `lora` transform. (And using a tracer would make it initial-style rather than final-style, but I don't think that difference is important.)

**Reply by davisyoshida** [reply]
Have you taken a look at https://github.com/google/jax/discussions/16280? I'd definitely appreciate people's thoughts on it.
